jetson/nucleo_bridge/test.cpp

```diff
@@ -0,0 +1,672 @@
+        else
+        {
+            printf("test adjust transaction successful on slave %i \n\n", addr);
+        }
+    }
+    catch (IOFailure &e)
+    {
+        printf("test adjust transaction failed on slave %i \n\n", addr);
+    }    
+}
+
+//tests
+// test functions for all slaves --> transactions are successful
+// full system test: off --> running --> off
+// good quad value test (i.e we run in and the value changes)
+// test that we dont set over max
+
+void testOff()
+{
+    PRINT_TEST_START
+    for (auto address : i2c_address)
+    {
+        // test off
+        off(address);
+        sleep(10);
+    }
+    PRINT_TEST_END
+}
+
+void testOn()
+{
+    PRINT_TEST_START
+    for (auto address : i2c_address)
+    {
+        // test off
+        on(address);
+        sleep(10);
+    }
+    PRINT_TEST_END
+}
+
+void testConfigPWM()
+{
+    PRINT_TEST_START
+    for (auto address : i2c_address)
+    {
+        // test off
+        int max = 30;
+        if (address < 17)
+        {
+            max = 16;
+        }
+        if (address == 16)
+        {
+            max = 16;
+        }
+        configPWM(address, max);
+
+        std::this_thread::sleep_for(std::chrono::milliseconds(10));
+    }
+    PRINT_TEST_END
+}
+
+void testQuadEnc()
+{
+    for (auto address : i2c_address)
+    {
+        // test off
+        uint32_t raw_angle = quadEnc(address);
+        printf("[%i] raw quad angle: %i \n", address, raw_angle);
+        std::this_thread::sleep_for(std::chrono::milliseconds(50));
+    }
+}
+
+void testAbsEnc()
+{
+    PRINT_TEST_START
+    for (auto address : i2c_address)
+    {
+        // test off
+        float raw_angle = static_cast<float>(absEnc(address));
+        printf("[%i] raw abs angle: %f \n", address, raw_angle);
+        sleep(10);
+    }
+    PRINT_TEST_END
+}
+
+void testClosed()
+{
+    PRINT_TEST_START
+    for (auto address : i2c_address)
+    {
+        int joint = (address & 0b1) + (((address >> 4) - 1) * 2); 
+        float p = 0.001; 
+        float i  = 0.00;
+        float d = 0; 
+
+        if (joint == 1)
+        {
+            p = 1.0;
+        }
+
+        setKPID(address, p * invert[joint], i * invert[joint], d * invert[joint]);
+        printf("joint %i, kp %f, ki, %f \n", joint, p * invert[joint], i * invert[joint] );
+        sleep(10);
+    }
+    while (1)
+    {
+        for (auto address : i2c_address)
+        {
+            int joint = (address & 0b1) + (((address >> 4) - 1) * 2); 
+            float angle = quadEnc(address);
+            
+            float target = angle - 0.15;
+            if (joint == 1)
+            {
+                angle = absEnc(address);
+                target = angle - 0.15;
+            }
+
+            do 
+            {
+                angle = closedPlus(address, target);
+                sleep(20);
+            }
+            while( joint != 1? abs(angle - target) > 0.01 : abs(angle - target) > 0.05);
+	        printf("arrived at position 1\n");
```

Comment:
Why are these lines indented throughout?

Reply:
unit tests, these aren't critical code that even runs on the rover

Reply:
this is a do while loop, the part in the do is indented